contracts/minter-suite/Minters/MinterPolyptych/MinterPolyptychV0.sol

```diff
@@ -0,0 +1,704 @@
+        );
+
+        // require artist to have configured price of token on this minter
+        require(_projectConfig.priceIsConfigured, "Price not configured");
+
+        // require token used to claim to be in set of allowlisted NFTs
+        require(
+            isAllowlistedNFT(_projectId, _ownedNFTTokenId),
+            "Only allowlisted NFTs"
+        );
+
+        // NOTE: delegate-vault handling **begins here**.
+
+        // handle that the vault may be either the `msg.sender` in the case
+        // that there is not a true vault, or may be `_vault` if one is
+        // provided explicitly (and it is valid).
+        address vault = msg.sender;
+        if (_vault != address(0)) {
+            // If a vault is provided, it must be valid, otherwise throw rather
+            // than optimistically-minting with original `msg.sender`.
+            // Note, we do not check `checkDelegateForAll` or `checkDelegateForContract` as well,
+            // as they are known to be implicitly checked by calling `checkDelegateForToken`.
+            bool isValidVault = delegationRegistryContract
+                .checkDelegateForToken(
+                    msg.sender, // delegate
+                    _vault, // vault
+                    genArt721CoreAddress, // contract
+                    _ownedNFTTokenId // tokenId
+                );
+            require(isValidVault, "Invalid delegate-vault pairing");
+            vault = _vault;
+        }
+
+        // we need the new token ID in advance of the randomizer setting a token hash
+        (uint256 _invocations, , , , , ) = genArtCoreContract.projectStateData(
+            _projectId
+        );
+        uint256 _newTokenId = (_projectId * ONE_MILLION) + _invocations;
+
+        // EFFECTS
+
+        // we need to store the new token ID before it is minted so the randomizer can query it
+        IGenArt721CoreContractV3WithRandomizer genArtCoreContractWithRandomizer = IGenArt721CoreContractV3WithRandomizer(genArt721CoreAddress);
+        bytes12 _hashSeed = genArtCoreContractWithRandomizer.tokenIdToHashSeed(
+            _ownedNFTTokenId
+        );
+        require(_hashSeed != 0);
+
+        genArtCoreContractWithRandomizer.randomizerContract()
+            .setPolyptychHashSeed(_newTokenId, _hashSeed);
+
+        // once mint() is called, the polyptych randomizer will either:
+        // 1) assign a random token hash
+        // 2) if configured, obtain the token hash from the `polyptychSeedHashes` mapping
+        tokenId = minterFilter.mint(_to, _projectId, vault);
+
+        // NOTE: delegate-vault handling **ends here**.
+
+        // okay if this underflows because if statement will always eval false.
+        // this is only for gas optimization (core enforces maxInvocations).
+        unchecked {
+            if (tokenId % ONE_MILLION == _projectConfig.maxInvocations - 1) {
+                _projectConfig.maxHasBeenInvoked = true;
+            }
+        }
+
```

Comment:
I think we need to have a per-project mapping of "redeemedTokenIds" that is used to limit each token to only be able to be used once to mint a polyptych token. For example, I shouldn't be able to mint 5 tokens with the same hash.

Reply:
Need to think more about this, I'm not sure that we need to introduce this (for now), it is somewhat of a subjective thing as to whether or not we _need_ this (depends on the intended behavior), and given that the actual minting/vouchering of previous mints was going to be handled via off-chain mechanisms given in-person minting, it probably is superfluous to add this check, but it does add a bit of implementation complexity.

That said, the implementation complexity it adds is pretty small, so maybe we do add it just as a safety rail (what you're describing is aligned with the initial planned usage of this minter).

Reply:
Good point, doesn't 100% need, as this could be handled via admin. I think it would be a nice safety rail, even if admin is running this, as it would protect us against any accidental duplicate mint transactions, etc., which tend to happen if there is a misunderstanding of blockchain state or something.

As discussed elsewhere, if this is not introduced, it does reduce complexity on the randomizer it we elect to guard the `setPolyptychHashSeed`, which we will likely do.

Reply:
@jakerockland - I talked to @ryley-o about this earlier today - here's the use case we should consider carefully:

- The artist wants to use the same project ID for a project that has three or more panels
- The tokens of the first panel are always minted out before the second panel is minted
- The second panel will share its token hash seed with the first panel, but in order to mint the third panel, we would need to allow that hash seed to be used more than once

If we want to support a polyptych within the same project (panels that share a project ID), we wouldn't be able to limit the original token's hash seed to a single reuse.

Reply:
Hrm–yeah this makes sense. Thank you both for digging more and re-flagging.

I do think that even if it is superfluous in the diptych framing, having this support from the outset here for gating hash-reuse could be the type of safety mechanism that is just worth the added complexity to introduce from the outset.

WDYT @ryley-o @shantanubala

Reply:
Some discussion ongoing in Slack, but perhaps a relatively minimal solution to prevent duplicate hashes on a given project is presented here: https://github.com/ArtBlocks/artblocks-contracts/pull/442/files